src/main/java/calculator/StringAddCalculator.java

```diff
@@ -0,0 +1,62 @@
+package calculator;
+
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+public class StringAddCalculator {
+    private static final String DEFAULT_SEPARATOR = "[,|:]";
+    private static final String CUSTOM_SEPARATOR = "//(.)\n(.*)";
+    private static final int DELIMITER_INDEX = 1;
+    private static final int TOKEN_INDEX = 2;
+
+    public static int splitAndSum(String text) {
+        int validateRsltFlag;
+        int totalSum = 0;
+
+        validateRsltFlag = validate(text);
+        if (validateRsltFlag >= 0) {
+            return validateRsltFlag;
+        }
+
+        validateNegativeNumberAndThrow(text);
+        
+        String[] numbers = splitBySeparator(text);
+        for (String num : numbers) {
+            totalSum += Integer.parseInt(num);
+        }
+
+        return totalSum;
+    }
```

Comment:
현재 splitAndSum이라는 함수는 인자를 분리하고 더한다는 의미를 가지고 있는데요! 
하나의 함수에서 최소 두 개의 기능이 구현되야함을 알 수 있습니다. 

그럼 테스트를  할 때 두 기능의 경우의 수를 모두 고려해서 테스트를 작성해야하는데, 2개까진 문제가 안될 수 있지만, 
3개, 4개가 될수록 테스트는 힘들어지겠죠! 왜 힘든지 잘 생각해보면 좋을 것 같아요. 

하나의 함수에서 많은 책임을 지지 않도록 하면 좋을 것 같습니다. 

우선 분리할 수 있는 대표적인 기능들을 구분해보면 
1. 인자 유효성 검증 
2. 인자 분리
3. 분리된 인자 총합 계산

이렇게 3가지인데요. 수도 코드로 구현해보면

```suggestion
    public static int splitAndSum(String text) {
        if(유효성검사(text) {
            // throw logic...
        }
        
        List<String> splitTextList = 문자열분리(text);
        return 총합계산(splitTextList);
    }
```

이렇게 나뉠 수 있겠죠!

Reply:
한솔님, 피드백 주신 수도코드를 바탕으로 다시 수정해보았어요.

제가 처음 코드작성시 '인자 유효성 검증'에 여러 검증이 있을 것 같다는 생각에 
큰 하나의 validate() 함수로 묶고 그 안에서 
1. null , empty 검증, 
2. 음수검증
3. 단일 문자열 검증 
이렇게 작성을 했었는데요

splitAndSum() 함수 내부에 제가 묶었던 검증 로직(validate())이 더 많은 책임을 지도록 한다는 말씀이실까요?

"그럼 테스트를 할 때 두 기능의 경우의 수를 모두 고려해서 테스트를 작성해야하는데, 2개까진 문제가 안될 수 있지만,
3개, 4개가 될수록 테스트는 힘들어지겠죠! 왜 힘든지 잘 생각해보면 좋을 것 같아요."
-> 책임을 많이 지면 안된다. 책임 분리원칙 까지는 이해가 가는데 테스트가 왜 힘들어지는지는  테스트코드에서 기능별로 메소드로 분리하면 되지 않을까 생각이 들어서 잘 떠오르지 않아서요.. 혹시 예시같은게 있을까요?

상세한 답변 감사드립니다 !

Reply:
@westkitekim 
음 예를 들어, 나는 분리가 제대로 되는지 여러 조건으로 테스트를 해야하는데, 내부적으로 분리뿐아니라 합계까지 같이 동작을 하다보니 내가 의도하는 결과가 나오는지 찾기가 힘들 수 있겠죠. 혹은 functionA라는 함수가 내부적으로 A라는 매개변수를 A-B-C-D로 데이터를 가공해서 D를 반환하는데, 내가 테스트하려는건 A에서 B로 변환하는게 제대로 되는지를 테스트하고 싶은 경우 하나의 함수로 이 부분을 테스트하기는 쉽지 않겠죠!